### ISSUE TITLE
Event when refresh fails

### ISSUE THREAD

**Issue by Dismissile** [IdentityModel/IdentityModel.AspNetCore#209]
When using services.AddAccessTokenManagement(), how can you respond to when getting a new access token fails? If the refresh token is no longer valid I need to be able to force a login again.

**Comment by VictorioBerra** [comment]
Probably have to implement you own IUserAccessTokenManagementService and swap it for the built in one:

https://github.com/IdentityModel/IdentityModel.AspNetCore/blob/main/src/AccessTokenManagement/UserAccessToken/UserAccessTokenManagementService.cs

Look at ` RefreshUserAccessTokenAsync` where if it fails to refresh, it just logs.

**Comment by natelaff** [comment]
Yes I think I am hitting this as well. When trying to come back to a web app that has this implemented sometimes I will get a 401. Researching IS4 logs indicates the refresh token has already been consumed and I end up DOA.

**Comment by natelaff** [comment]
Yes I think I am hitting this as well. When trying to come back to a web app that has this implemented sometimes I will get a 401. Researching IS4 logs indicates the refresh token has already been consumed.

**Comment by leastprivilege** [comment]
I thought about this. But ultimately, getting a 401 is your event. If the library cannot provide you with a "working" access token, then all you can do is start over and do an authorize request.

Not sure what the value would be of some event being triggered on some backgrund thread...

**Comment by Dismissile** [comment]
I'm curious what a good approach would this to be. If we use the built in helper that adds a delegating handler to the HttpClient then we can't tell if there is an access token before it makes the call to a downstream API.  I could check the response of the HTTP call for a 401 and then do a challenge?

I'm guessing that doing a challenge on the background thread would not work and that is why you're saying it's not very useful?

**Comment by leastprivilege** [comment]
If you get back a 401 from an API call using this library - then the library did everything it could to give you a valid token.

You need to start over doing a challenge.

**Comment by natelaff** [comment]
Digging this back up.

> If you get back a 401 from an API call using this library - then the library did everything it could to give you a valid token.
> 
> You need to start over doing a challenge.

Agreed on the 401-- but, what about the cases where the library fails to retrieve a token in the first place? Where would be the best place to check for this as is? We have to supply our own TokenEndpointService? I think an event firing would actually be quite useful on a 400 failure on refresh token failed.

Example, token already consumed errors. We know we'll get a 400 back, we can challenge again, but can't figure out the best place to hook into these.

**Comment by leastprivilege** [comment]
So what would you do when this event fires?

**Comment by natelaff** [comment]
Throw a more specific (custom) exception that could be caught in pipeline to trigger a challenge? That would work.... right?

**Comment by leastprivilege** [comment]
So IOW I could throw an exception when a 400 is returned?

**Comment by leastprivilege** [comment]
Also - if no token is returned - isn't that the same? I think we return null if we cannot provide a current token

**Comment by natelaff** [comment]
Yes, if no token is returned that's the same. But in that event I find myself in the same position as before. Where to hook into this to say "ok, token no longer exists, I need to challenge" ASP.NET Core web app isn't doing that.

**Comment by leastprivilege** [comment]
We discussed this internally - we don't think that this event will give you what you want - if you would start a challenge in this event, still your main code will probably step over the challenge by returning some data etc.

You need to check for 401 at the call site. If you get a 401 and our library cannot resolve this - then you need to challenge in your business code. Or add another message handler that checks for the 401 and throw an exception.

**Comment by natelaff** [comment]
Understood :)

**Comment by agrisyst-slo** [comment]
I wish I would get a 401 result when the refresh token is no longer valid (expired). 
As @VictorioBerra said: Look at RefreshUserAccessTokenAsync where if it fails to refresh, it just logs. 
The access token is returned as is, not null.

I have read all the documentation several times and tried a lot of different configurations, but expired refresh tokens do not result in a 401 status code and a redirect to the login page.

This is the configuration for the refresh token:
RefreshTokenUsage = TokenUsage.OneTimeOnly,
RefreshTokenExpiration = TokenExpiration.Absolute,
AbsoluteRefreshTokenLifetime = 65, // For testing

This is some logging which shows that the same refresh token is used several times:
[09:47:54 DBG] Token for user saskia needs refreshing.
[09:47:54 DBG] Refreshing refresh token: 822B1EC171890410EF7868BDCC675FAEEBA00FBDC07F02A38A86F25FA6F8FFF5
[09:49:02 DBG] Token for user saskia needs refreshing.
[09:49:02 DBG] Refreshing refresh token: 865662EEC7734B6D766B273B121EB57EB4A1EB1EB32F2093BC38AFD28EBDA87A
[09:49:02 ERR] Error refreshing access token. Error = invalid_grant
[09:49:02 DBG] Token for user saskia needs refreshing.
[09:49:02 DBG] Refreshing refresh token: 865662EEC7734B6D766B273B121EB57EB4A1EB1EB32F2093BC38AFD28EBDA87A
[09:49:02 ERR] Error refreshing access token. Error = invalid_grant

I looks like a bug to me. If not, please let me know how to get the 401 for a expired refresh token.

**Comment by leastprivilege** [comment]
Need to look into it when I have time. In the meantime feel free to dig deeper.

**Comment by leastprivilege** [comment]
I tried to repro it with the Web sample from the repo.

1. login
2. call API as user
3. go to demo token server and revoke token
4. call API again
5. 401

<img width="1020" alt="image" src="https://user-images.githubusercontent.com/1454075/152760334-26da31ab-43bc-4983-9971-d6b910c0b117.png">

**Comment by agrisyst-slo** [comment]
The problem happens when the refresh token is expired (without explicitly revoking it).
I will create a simple demo app that shows the problem as soon as possible.

**Comment by leastprivilege** [comment]
> I will create a simple demo app

You don't have to do that. Rather suggest a way how we should handle that.

**Comment by leastprivilege** [comment]
IOW - you want to tie the lifetime of the refresh token to the session lifetime?

**Comment by agrisyst-slo** [comment]
There is some miscommunication here. I don't know how to make myself understood, but I will try again.

I thought that this option takes care of tying the session and token lifetime together:
options.UseTokenLifetime = true;
If I misunderstood, please let me know. Even though I don't think that that is my problem.

As to the refresh token, I thought that the access token gets refreshed until the refresh token is expired. After that a user must enter his credentials again.
If that is not the case, exactly when does a user have to renew his credentials?
And what is the purpose of the refresh token, if not to refresh the access token either until the refresh token is expired (fixed time) or until the last refresh was too long ago (sliding refresh).

And what should happen in the case a refresh token is expired? I would expect the access token to be no longer valid. If that is not the case, how can I find out that the access token is invalid? That is, without the 401 response that I for some reason don't get. And that is my problem (I think). For some reason I do not get a 401 response when the refresh token and therefore the access token is expired.

I have seen examples where the expiration time of the token is checked to see whether the access token must be refreshed etc. But I thought that the IdentityModel.AspNetCore library takes care of that. If I misunderstood, please let me know.

**Comment by leastprivilege** [comment]
> options.UseTokenLifetime = true

Is used to tie the session lifetime to the lifetime of the identity token. Not the access token.

All I can say is - in my testing. Regardless why the refresh token is not valid anymore (revoked or expired), **the next time you will try to use the access token**, the API call will return a 401. This needs to be handled at the app level.

**Comment by leastprivilege** [comment]
Maybe you need to debug the app/library and repro the problem to figure out what's going on?

**Comment by agrisyst-slo** [comment]
I will start from scratch again and see if I can get the 401 after the refresh token expires.
If I find out what the problem is, I will let you know. Maybe that can help somebody else.
Thank you for your help.

**Comment by leastprivilege** [comment]
any update?

**Comment by agrisyst-slo** [comment]
At the moment I am very busy with another project.
For now my not so elegant work around will have to do for the project that uses the identity server.
As soon as I have the time I will try to solve the problem and let you know if I have figured it out.
